### PR TITLE
Remove deprecated Lookup#resolve usage from Rule matching code

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
FYI: @ArturGajowy I think you might be interested ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java [hunk 1]
```diff
@@ -140,16 +145,43 @@ private Patterns() {}
...
+                            .collect(toImmutableList());
+
+                    return Streams.stream(new AbstractIterator<List<PlanNode>>() {
+                        @Override
+                        protected List<PlanNode> computeNext()
+                        {
+                            List<PlanNode> sources = sourceStreams.stream()
+                                    .map(source -> source.findFirst()
+                                            .orElse(null)).collect(toImmutableList());
+                            sourceStreams.stream().forEach(sourceStream -> sourceStream.skip(1));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java]: Did you mean to discard the effect of `sourceStream.skip(1)`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java [hunk 2]
```diff
@@ -140,16 +145,43 @@ private Patterns() {}
...
+                    List<Stream<PlanNode>> sourceStreams = node.getSources().stream()
+                            .map(source -> context.getLookup().resolveGroup(source))
+                            .collect(toImmutableList());
+
+                    return Streams.stream(new AbstractIterator<List<PlanNode>>() {
+                        @Override
+                        protected List<PlanNode> computeNext()
+                        {
+                            List<PlanNode> sources = sourceStreams.stream()
+                                    .map(source -> source.findFirst()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java]: Once you call `findFirst`, the stream should not be used anymore.
Maybe convert to `List<Iterator>`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java [hunk 3]
```diff
@@ -140,16 +145,43 @@ private Patterns() {}
...
+                                            .orElse(null)).collect(toImmutableList());
+                            sourceStreams.stream().forEach(sourceStream -> sourceStream.skip(1));
+
+                            if (sources.contains(null)) {
+                                return endOfData();
+                            }
+                            return sources;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java]: You're basically making a matrix of possible source combinations  and transforming it (matrix-wise).

1. can you simply enumerate the possibilities and make this transformation simpler?
2. you arbitrarily combine first member in each group as one combination, second member of each group as another combination, etc... Is it correct, without making a cartesian product here? (which could obviously explode)
3. .. which boils down to conclusion: should you do this at all without actually changing iterative's logic? This starts to outgrow iterative optimizer's assumption

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 4]
```diff
@@ -157,7 +158,7 @@ private RuleApplication applyRule()
 
     private static <T> RuleApplication applyRule(Rule<T> rule, PlanNode planNode, Rule.Context context)
     {
-        PlanNodeMatcher matcher = new PlanNodeMatcher(context.getLookup());
+        Matcher matcher = new DefaultMatcher();
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: `DefaultMatcher` should be gone too :)

**Comment by kokosing** [comment]
I extracted simple commits (those which are not related to `ExplorePattern`) here #9787

**Comment by ArturGajowy** [comment]
Hi @kokosing! Are there plans to work on this? Or can we close it?

**Comment by kokosing** [comment]
> Hi @kokosing! Are there plans to work on this? Or can we close it?

Yes, there are. But with low priority.

**Pull request closed without merging**